components/scream/src/physics/shoc/shoc_calc_shoc_vertflux_impl.hpp

```diff
@@ -0,0 +1,37 @@
+#ifndef SHOC_CALC_SHOC_VERTFLUX_IMPL_HPP
+#define SHOC_CALC_SHOC_VERTFLUX_IMPL_HPP
+
+#include "shoc_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace shoc {
+
+template<typename S, typename D>
+void Functions<S,D>
+::calc_shoc_vertflux(
+  const MemberType& team,
+  const Int& nlev,
+  const uview_1d<const Spack>& tkh_zi,
+  const uview_1d<const Spack>& dz_zi,
+  const uview_1d<const Spack>& invar,
+  const uview_1d<Spack>& vertflux)
+{
+  const Int nlev_pack = scream::pack::npack<Spack>(nlev);
+  const auto sinvar = scalarize(invar);
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team, nlev_pack), [&] (const Int& k) {
+    auto range_pack1 = scream::pack::range<IntSmallPack>(k*Spack::n);
+    auto range_pack2 = range_pack1;
+    range_pack2.set(range_pack1 < 1, 1); // don't want the shift to go below zero. we mask out that result anyway
+
+    Spack up_grid, grid;
+    pack::index_and_shift<-1>(sinvar, range_pack2, grid, up_grid);
+    const Spack grid_dz = 1 / dz_zi(k); // vertical grid diff squared
+    // Compute the vertical flux via downgradient diffusion
+    vertflux(k).set(range_pack1 > 0 && range_pack1 < nlev, -(tkh_zi(k) * grid_dz * (up_grid - grid)));
```

Comment:
I'm not going do dive into the details of indexing, but just as a check: Are you sure `> 0` is right for 0-based C indexing vs 1-based F90? Same for `range_pack1 < 1` above.

Reply:
yes. if you look at the f90 implementation, it skips the first k index.